### PR TITLE
Add support for JSON logging.

### PR DESCRIPTION
**Pull request by youngnick**
Updates #624

This PR contains the code for JSON logging only. The ticket will be closed when the documentation is merged as well.

Signed-off-by: Nick Young <ynick@vmware.com>

### CODE REVIEW

File: internal/contour/listener.go [hunk 1]
```diff
@@ -127,6 +139,68 @@ func (lvc *ListenerVisitorConfig) httpsAccessLog() string {
...
+	return DEFAULT_ACCESS_LOG_TYPE
+}
+
+// accesslogFields returns the access log fields that should be configured
+// for Envoy, or a default set if not configured.
+func (lvc *ListenerVisitorConfig) accesslogFields() []string {
+	if lvc.AccessLogFields != nil {
+		return lvc.AccessLogFields
+	}
+	return []string{
```
Comment (davecheney) [inline internal/contour/listener.go]: Please move this to a public variable next to the translation map. This is so we can reference it as the canonical source from the release notes.

File: internal/contour/listener.go [hunk 2]
```diff
@@ -127,6 +139,68 @@ func (lvc *ListenerVisitorConfig) httpsAccessLog() string {
...
+		"user_agent",
+		"x_forwarded_for",
+	}
+}
+
+func (lvc *ListenerVisitorConfig) newInsecureAccessLog() []*envoy_api_v2_accesslog.AccessLog {
+	var accesslogger []*envoy_api_v2_accesslog.AccessLog
+	switch lvc.accesslogType() {
+	case "json":
+		accesslogger = envoy.FileAccessLogJSON(lvc.httpAccessLog(), lvc.accesslogFields())
```
Comment (davecheney) [inline internal/contour/listener.go]: case "json":
      return envoy.FileAccessLogJson...
default:
     return envoy.FileAccessLog...
}

File: internal/contour/listener.go [hunk 3]
```diff
@@ -127,6 +139,68 @@ func (lvc *ListenerVisitorConfig) httpsAccessLog() string {
...
+	default:
+		accesslogger = envoy.FileAccessLog(lvc.httpAccessLog())
+	}
+	return accesslogger
+}
+
+func (lvc *ListenerVisitorConfig) newSecureAccessLog() []*envoy_api_v2_accesslog.AccessLog {
+	var accesslogger []*envoy_api_v2_accesslog.AccessLog
+	switch lvc.accesslogType() {
+	case "json":
```
Comment (davecheney) [inline internal/contour/listener.go]: same as above

File: internal/contour/listener.go [hunk 4]
```diff
@@ -209,6 +285,7 @@ func (l listenersByName) Less(i, j int) bool {
 	return l[i].(*v2.Listener).Name < l[j].(*v2.Listener).Name
 }
 
+// TypeURL returns the cache's Listener Type.
```
Comment (davecheney) [inline internal/contour/listener.go]: nit: this is why I don't pay attention to the lint warnings ... this style of comment adds little beyond shutting the linter up.

File: internal/contour/listener.go (hunk 2 above)
Comment (youngnick) [reply internal/contour/listener.go]: I wasn't sure if that style was okay, thanks.

File: internal/contour/listener.go (hunk 4 above)
Comment (youngnick) [reply internal/contour/listener.go]: agreed, removed.

**Pull request merged**